### DISCUSSION TITLE
Way to find out which program is installed?

### DISCUSSION THREAD

**Discussion by rbeckett75** [pybricks/support#1604] [Ideas]
I think I mentioned before, that it would be nice to be able to read a program back from a hub - but fully aware that it's not possible due to the translation from python to machine code..

BUT... would it be possible for pybricks to at identify the filename, or some program name that is currently stored on the hub?

I have quite a few hubs.. and sometimes they don't get used for a while, and it would be handy to be able to connect to it, and pybricks could tell me that the last program downloaded was "Motor_Control_Crane_v1.5.py"  or something

Also - are you looking into supporting multiple programs on inventor/spike hubs?  ala the way the official Lego software works?

thanks again for all the great work! :)

**Comment by laurensvalk** [comment]
It's a bit of a workaround, but may do the job: You might know that you can read data that you have stored yourself. But you can read beyond that, which gives you the compiled user program. Run the following on the REPL (`>>>` button), otherwise you'd replace your old program with the new.

```
USER_STORAGE = 512; # For prime/essential hub, 128 for all other hubs
hub.system.storage(USER_STORAGE + 17, read=40)
```

So it reads starting from the end of user storage, plus 17 to skip some header information for convenience. I chose 40 bytes to read, but just pick something higher if your program name is not visible in the output.

On my hub I get: 

```
>>> USER_STORAGE = 512; hub.system.storage(USER_STORAGE + 17, read=40)
b'M\x06\x00\x1f\x13\x01\x10HELLO.py\x00\x0f\x16hello_world\x00*pybricks.'
```

So the script on the hub was `HELLO.py`.

> Also - are you looking into supporting multiple programs on inventor/spike hubs? ala the way the official Lego software works?

See the [`hub_menu`](https://docs.pybricks.com/en/latest/tools/index.html#pybricks.tools.hub_menu) for an example.

It's not entirely the same, but gives you a lot of flexibility. You could even use this technique on the small hubs with one button, if you make another way to do the selection. For example by cycling through several colors with the button, or based on sensor input, etc.

**Comment by BertLindeman** [comment]
Which program was loaded onto one of my Technic hubs?
Laurens procedure above did not help me enough as I use various firmware (beta) levels. 😢 

Upcoming changes wil change the location more, I think.

## REPL
Use the REPL to keep the loaded program unchanged!

Paste the complete program including the empty lines (they end the while statements)
or paste line by line, as you wish.
Paste the in the REPL of the hub you want to know the name of the program of (Obviously).

```python
startloc = 0
look_for_6 = b'\x00M\x06\x00'
look_for_6_1 = b'\x00M\x06\x01'
while hub.system.storage(startloc, read=4) not in [look_for_6, look_for_6_1]: startloc += 1

startloc += 1
print("startloc", startloc, "\n", hub.system.storage(startloc, read=60))
fn = bytearray(); i = 0; byte = hub.system.storage(startloc + 7 + i, read=1); 
while byte != b'\x00': fn.extend(byte); i += 1; byte = hub.system.storage(startloc + 7 + i, read=1)

filename_str = str(fn, 'utf-8')
print("\nfilename:", filename_str)
```

This procedure is full of assumptions, so no guarantee.

## Explanation:
- Starting to search at location zero for the Micropython mpy header.
The mpy header indicates the start of the loaded mpy file. 
Start at zero is very exagerated, 512 will be a good start.
- mpy header is the start of the mpy file, see below for the fields.
- Current mpy version is 6.0. Just to be sure I search for 6.1 as well.
- I assume there always is a hex zero before the mpy header so if found: add one to startloc
- 60 bytes after the start of the mpy header are printed.
Most of the time enough to puzzle which program is loaded.
- 7 bytes after startloc starts a null terminated string: the program file name.
No path is included. Well it **is** micropython.

If you run this program (so not in the REPL), just to test how it works, add these two lines on the top:

```python
from pybricks.hubs import ThisHub
hub = ThisHub()
```
They are not needed in the REPL as they are already done.

As example the result in the REPL:
```
>>> startloc = 0
>>> look_for_6 = b'\x00M\x06\x00'
>>> look_for_6_1 = b'\x00M\x06\x01'
>>> while hub.system.storage(startloc, read=4) not in [look_for_6, look_for_6_1]: startloc += 1
... 
>>> startloc += 1
>>> print("startloc", startloc, "\n", hub.system.storage(startloc, read=60))
startloc 637 
 b'M\x06\x00\x1fB\x11.print_program_on_hub.py\x00\x0f\x0eThisHub\x00\x1apybricks.hubs\x00\x0csys'
>>> fn = bytearray(); i = 0; byte = hub.system.storage(startloc + 7 + i, read=1); 
>>> while byte != b'\x00': fn.extend(byte); i += 1; byte = hub.system.storage(startloc + 7 + i, read=1)
... 
>>> filename_str = str(fn, 'utf-8')
>>> print("\nfilename:", filename_str)

filename: print_program_on_hub.py
>>> 
```

### mpy Header:
| size | field |
| -- | -- |
| byte | value 0x4d (ASCII ‘M’) |
| byte | .mpy major version number |
| byte | native arch and minor version number (was feature flags in older versions) |
| byte | number of bits in a small int |


Hope you never need it.

Fun puzzling this,

Bert
